botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1075 @@
+from functools import lru_cache
+from typing import NamedTuple
+
+from botocore import xform_name
+from botocore.compat import quote, urlparse
+from botocore.exceptions import (
+    EndpointInputParametersError,
+    EndpointResolutionError,
+    RuleSetConfigurationError,
+)
+from botocore.utils import (
+    ArnParser,
+    InvalidArnException,
+    is_valid_ipv4_endpoint_url,
+    is_valid_ipv6_endpoint_url,
+    normalize_url_path,
+    percent_encode,
+)
+
+logger = logging.getLogger(__name__)
+
+TEMPLATE_STRING_RE = re.compile(r'\{[a-zA-Z#]+\}')
+VALID_HOST_LABEL_RE = re.compile(
+    r"^((?!-)[A-Z\d-]{1,63}(?<!-)\.)*((?!-)[A-Z\d-]{1,63}(?<!-))$",
+    re.IGNORECASE,
+)
+GET_ATTR_RE = re.compile(r'(\w+)\[(\d+)\]')
+
+
+class BaseRule:
+    """A base class representing a rule within a rule set. All rules contain a
+    conditions property, which can be empty, a partitions JSON and documentation
+    about the rule.
+    """
+
+    def __init__(self, conditions, partitions, documentation=None):
+
+        self.conditions = conditions
+        self.partitions = partitions
+        self.documentation = documentation
+
+    def evaluate(self, input_parameters):
+        raise NotImplementedError()
+
+    def is_func(self, argument):
+        """Determine if an object is a function object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a function object
+
+        :returns: A bool indicating if the input is a function object
+        """
+        return isinstance(argument, dict) and 'fn' in argument
+
+    def is_ref(self, argument):
+        """Determine if an object is a reference object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a reference object
+
+        :returns: A bool indicating if the input is a reference object
+        """
+        return isinstance(argument, dict) and 'ref' in argument
+
+    def is_template(self, argument):
+        """Determine if an object contains a template string.
+
+        :type argument: Any
+        :param argument: An object that may or may not contain a template string.
+
+        :returns: A bool indicating if the input contains a template string.
+        """
+        return (
+            isinstance(argument, str)
+            and TEMPLATE_STRING_RE.search(argument) is not None
+        )
+
+    def resolve_template_string(self, value, input_parameters):
+        """Given a templated string, parse out the template nested between curly
+        braces, and return the nested data from the rule's input parameters.
+
+        :type value: str
+        :param value: A template string e.g. "https://{Bucket}.ec2.{url#authority}"
+
+        :returns: A fully resolved string e.g. "https://us-west-2.ec2.amazonaws.com"
+        """
+        matches = TEMPLATE_STRING_RE.findall(value)
+        for match in matches:
+            # first and last will be '{' and '}'
+            template_value = match[1:-1]
```

Comment:
I did that in an earlier version. However, there are parsing issues with f strings and double brackets in a string replace method. this was the 'cleanest' way to avoid that problem. 

i commented on another thread about this that this regex does work for double bracket template strings. however i haven't found it in the spec or in any generated test cases. i've reached out to the spec writers about this.

Reply:
I think what @nateprewitt means by "account for `{{Bucket}}`" is that `{{Bucket}}` should remain unchanged. And I think that this code would change `{{Bucket}}`:

```python
>>> import re
>>> TEMPLATE_STRING_RE = re.compile(r'\{[a-zA-Z#]+\}')
>>> matches = TEMPLATE_STRING_RE.finall()
>>> matches = TEMPLATE_STRING_RE.findall("abc {{Bucket}} {foo}")
>>> matches
['{Bucket}', '{foo}']
```

Using lookahead and lookbehind like below may address that, but may trip over other edge cases like `{{foo} bar}`:

```
(?<!\{)\{{1}[a-zA-Z#]+\}(?!\})
```

Reply:
Have you considered using [Python's built-in `string.format` tooling](https://docs.python.org/3/library/string.html#string.Formatter) instead of regex for parsing the template strings? I think it meets the spec and comes with all the edge cases taken care of (as validated by Python's own unit tests)

Reply:
I see. I'll try to figure out a way for the regex pattern to keep nested brackets in there. As for using `string.format`, that could be a great solution if not for nested attributes to the right of `#`. I still think regex would be preferable here if there is a simple way to not strip out brackets.

Reply:
All that was missing was explicitly stating the range of expected occurrences of brackets in the regex:
`\{{1,2}[a-zA-Z#]+\}{1,2}`
While it may seem more ideal to strip brackets out using lookaheads/behinds, they still need to be put back in when looking for the template to replace, so I don't really see any advantage in doing that

Reply:
To be clear, we don't want to perform substitution on `{{bucket}}`. The specification already lists this behavior, but to make sure it's cataloged here, double curly brackets is an escape for substitution. `{{bucket}}` is the string literal `{bucket}`. It should never be evaluated for templating.